### PR TITLE
Use batch_size without modulo

### PR DESCRIPTION
**Pull request by bhack**
Please check #44933

### CODE REVIEW

File: tensorflow/python/data/ops/dataset_ops.py [hunk 1]
```diff
@@ -3901,9 +3901,13 @@ def __init__(self, input_dataset, batch_size, drop_remainder):
     self._drop_remainder = ops.convert_to_tensor(
         drop_remainder, dtype=dtypes.bool, name="drop_remainder")
 
-    constant_drop_remainder = tensor_util.constant_value(self._drop_remainder)
     # pylint: disable=protected-access
-    if constant_drop_remainder:
+    try:
+      batch_modulo = self._input_dataset.__len__() % self._batch_size
```
Comment (aaudiber) [inline tensorflow/python/data/ops/dataset_ops.py:3906]: It's tricky how sometime batch_modulo is a number and sometimes a bool. Can we call it `evenly_divisible` and set it to `self._input_dataset.__len__() % self._batch_size == 0`?

File: tensorflow/python/data/ops/dataset_ops.py [hunk 2]
```diff
@@ -3901,9 +3901,13 @@ def __init__(self, input_dataset, batch_size, drop_remainder):
...
+    except TypeError:
+      batch_modulo = True
```
Comment (aaudiber) [inline tensorflow/python/data/ops/dataset_ops.py:3908]: Comment about why we would get `TypeError`

`# Indicates that the length of the dataset could not be determined`

**Review by aaudiber — CHANGES_REQUESTED** [review]
Thanks @bhack! Please also add a unit test in tensorflow/python/data/kernel_tests/batch_test.py

**Comment by bhack** [comment]
@aaudiber Mhh.. we are going to need to diverge some tests like this between eager and graph mode. 
So how much extra work do you think is needed to expand this patch over graph mode to maintain the tests unified?

https://github.com/tensorflow/tensorflow/blob/063940bae365534a74d27421c7ef37736042cb3a/tensorflow/python/data/experimental/kernel_tests/rebatch_dataset_test.py#L184-L191

**Comment by bhack** [comment]
I found a small free time slot for an initial refactoring to try to cover graph mode. Not all test are passing.

**Comment by bhack** [comment]
I've tried a quick hack with `tf.keras.backend.eval` for the failing distribution tests but what we could use for a for a replica safe evaluation?

**Review by aaudiber — APPROVED** [review]
Thanks @bhack

File: tensorflow/python/data/ops/dataset_ops.py [hunk 3]
```diff
@@ -454,6 +455,23 @@ def __len__(self):
...
 
+  def __evenly_divisible__(self, batch_size):
+    if not context.executing_eagerly():
+      batch_mod = math_ops.floor_mod(self.cardinality(),batch_size)
+      cond = math_ops.logical_and(math_ops.greater(self.cardinality(),0),
+                                  math_ops.equal(batch_mod,0))
+      pred = smart_cond.smart_cond(cond, lambda: constant_op.constant(True),
+                               lambda: constant_op.constant(False))
+      #TODO(bhack) improve this
+      evenly_divisible = backend.eval(pred)
```
Comment (bhack) [inline tensorflow/python/data/ops/dataset_ops.py:466]: @aaudiber How can I evaluate this removing Keras but also pass Distributed test?

**Comment by aaudiber** [comment]
> @aaudiber Mhh.. we are going to need to diverge some tests like this between eager and graph mode.
> So how much extra work do you think is needed to expand this patch over graph mode to maintain the tests unified?
> 
> https://github.com/tensorflow/tensorflow/blob/063940bae365534a74d27421c7ef37736042cb3a/tensorflow/python/data/experimental/kernel_tests/rebatch_dataset_test.py#L184-L191

It would be a lot of work since all cardinality logic is currently in C++. Given that we would need to diverge a lot of tests between eager and graph mode, I think it's better to wait until the logic is migrated to Python.

**Review by aaudiber — CHANGES_REQUESTED** [review]
Given the complexity this adds to testing, I don't think it makes sense to do this special casing for eager mode. Users can always work around the issue by setting `drop_remainder=True`.

**Comment by bhack** [comment]
I have no more special testing needs and I suppose it could work in graph mode.

The issue now how to safely evaluate `evenly_divisible`.

**Comment by bhack** [comment]
@aaudiber Can you re-check now? I've not diverged the test between edge and graph mode.
Now is rollback on an early formulation when the only failing test is `testMultipleBatches` in graph_mode.

**Comment by bhack** [comment]
The single test is failing cause if we extract a simple example:

```python
import tensorflow as tf
from tensorflow.python.framework import tensor_util
from tensorflow.python.client import session as session_lib

ds = tf.data.Dataset.range(16)
print("Card eager. ", ds.cardinality().numpy())
tf.compat.v1.disable_eager_execution()
ds = tf.data.Dataset.range(16)
print("Card const: ", tensor_util.constant_value(ds.cardinality()))
with session_lib.Session() as session:
    print("Card eval: ", ds.cardinality().eval())
```

```
Card eager.  16
Card const:  None
Card eval:  16
```

It cannot extract the value from `constant_value`. How can I safely evaluate?

**Comment by aaudiber** [comment]
@bhack this looks like a fundamental issue in graph mode. We can't get the result of `cardinality` because it requires evaluating the cardinality op, which can't be done eagerly when in graph mode. The only way I can think of is to implement cardinality for each type of dataset as a Python method, so that the method can produce the right cardinality in both graph mode and eager mode.

**Comment by bhack** [comment]
> @bhack this looks like a fundamental issue in graph mode. We can't get the result of `cardinality` because it requires evaluating the cardinality op, which can't be done eagerly when in graph mode. The only way I can think of is to implement cardinality for each type of dataset as a Python method, so that the method can produce the right cardinality in both graph mode and eager mode.

Why it  happens only for a single test? All the other tests are passing. Is it only a problem for `RangeDataset`?

**Comment by bhack** [comment]
@aaudiber I made just an extra commit to show that all the test are passing.

Is it possible that the problem of cardinality occurs in a single test?
Is it a coincidence?

**Comment by bhack** [comment]
I've fixed also some GPU and distributed test It was a little bit slow cause currently I don't have a spare GPU for executing this tests so I needed to use the public CI. .
Now all the tests are passing also on GPU.

I'am only diverging a single `def` test for legacy rebatch for graph mode.

It will need a refactor for sure but lets give a quick review pass.

**Comment by bhack** [comment]
I've reverted the last refactoring commit https://github.com/tensorflow/tensorflow/pull/45047/commits/f8466c83fa6f5c77fc9203b7dbd4042f1461d08e cause it was creating some strange error on function wrapping in edge mode. Tell me if you have any feedback as the old build is still available

**Comment by bhack** [comment]
Let me know.

File: tensorflow/python/data/ops/dataset_ops.py [hunk 4]
```diff
@@ -3901,9 +3909,10 @@ def __init__(self, input_dataset, batch_size, drop_remainder):
     self._drop_remainder = ops.convert_to_tensor(
         drop_remainder, dtype=dtypes.bool, name="drop_remainder")
 
-    constant_drop_remainder = tensor_util.constant_value(self._drop_remainder)
     # pylint: disable=protected-access
-    if constant_drop_remainder:
+    evenly_divisible = input_dataset.__evenly_divisible__(self._batch_size)
+    constant_drop_remainder = tensor_util.constant_value(self._drop_remainder)
+    if constant_drop_remainder or tensor_util.constant_value(evenly_divisible):
```
Comment (aaudiber) [inline tensorflow/python/data/ops/dataset_ops.py:3915]: In graph mode will `tensor_util.constant_value(evenly_divisible)` ever return something other than `None`?

**Review by aaudiber — CHANGES_REQUESTED** [review]

File: tensorflow/python/data/ops/dataset_ops.py (hunk 4 above)
Comment (bhack) [reply tensorflow/python/data/ops/dataset_ops.py:3915]: It was quite one month old and I need to recompile everything cause I've not RBE access as the TF team and I was on others PR/branches. I've reverted a single test change to see if it fail in graph mode.

Comment (bhack) [reply tensorflow/python/data/ops/dataset_ops.py:3915]: Ok it seems It has passed a special test in graph mode only so probably the original tests protfolio can pass with the none case in graph mode as all the tests passed.

As I don't have the cache ready for this PR please close this if you don't see any other solution to have exactly the same behaviour in graph mode with drop false.

**Pull request closed without merging**